hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java

```diff
@@ -0,0 +1,260 @@
+
+    private Deflater deflater;
+
+    private int headerOff = 0;
+
+    private byte[] userBuf = null;
+    private int userBufOff = 0;
+    private int userBufLen = 0;
+
+    private int headerBytesWritten = 0;
+    private int trailerBytesWritten = 0;
+
+    private int currentInputLen = 0;
+
+    private Checksum crc = DataChecksum.newCrc32();
+
+    private BuiltInGzipDecompressor.GzipStateLabel state;
+
+    public BuiltInGzipCompressor(Configuration conf) {
+        ZlibCompressor.CompressionLevel level = ZlibFactory.getCompressionLevel(conf);
+        ZlibCompressor.CompressionStrategy strategy = ZlibFactory.getCompressionStrategy(conf);
+
+        // 'true' (nowrap) => Deflater will handle raw deflate stream only
+        deflater = new Deflater(level.compressionLevel(), true);
+        deflater.setStrategy(strategy.compressionStrategy());
+
+        state = BuiltInGzipDecompressor.GzipStateLabel.HEADER_BASIC;
+        crc.reset();
+    }
+
+    @Override
+    public boolean finished() {
+        return deflater.finished();
+    }
+
+    @Override
+    public boolean needsInput() {
+        if (state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+            return deflater.needsInput();
+        }
+
+        return (state != BuiltInGzipDecompressor.GzipStateLabel.FINISHED);
+    }
+
+    @Override
+    public int compress(byte[] b, int off, int len) throws IOException {
+        int numAvailBytes = 0;
+
+        // If we are not within uncompression data yet. Output the header.
+        if (state != BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+            if (userBufLen <= 0) {
+                return numAvailBytes;
+            }
+
+            int outputHeaderSize = writeHeader(b, off, len);
+
+            // Completes header output.
+            if (headerOff == 10) {
+                state = BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM;
+            }
+
+            numAvailBytes += outputHeaderSize;
+
+            if (outputHeaderSize == len) {
+                return numAvailBytes;
+            }
+
+            off += outputHeaderSize;
+            len -= outputHeaderSize;
+        }
+
+        if (state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+            // hand off user data (or what's left of it) to Deflater--but note that
+            // Deflater may not have consumed all of previous bufferload, in which case
+            // userBufLen will be zero
+            if (userBufLen > 0) {
+                deflater.setInput(userBuf, userBufOff, userBufLen);
+
+                crc.update(userBuf, userBufOff, userBufLen);  // CRC-32 is on uncompressed data
+
+                currentInputLen = userBufLen;
+                userBufOff += userBufLen;
+                userBufLen = 0;
+            }
+
+
+            // now compress it into b[]
+            int deflated = deflater.deflate(b, off, len - 8, Deflater.FULL_FLUSH);
```

Comment:
Need to reserve for the trailer. Otherwise if deflater reported finished status, we have not chance to output trailer.

Hmm, I set it to FULL_FLUSH to make sure `compress` will output all compressed output immediately. Just to make sure the compression work without other factor I'm not aware.

As now all tests are passed, we can test with NO_FLUSH and see if it works.

Reply:
hmm what if `len < 8`, will it run into error?

Reply:
Tried with NO_FLUSH locally and looks passing tests. Let's see Hadoop CI.

Reply:
> hmm what if `len < 8`, will it run into error?

If buffer size < 0, `deflate` will throw `ArrayIndexOutOfBoundsException`, we should prevent this case.

If it is 0, `deflate` won't write compressed data out.

Reply:
Hmm, you remind me that in edge case, e.g. the buffer size is less than 8, we still need to be able to compress. The trailer output should be able to handle the case. I will update it.

Reply:
Yea, I think you also need to count the header size so it could happen when the buffer size >= 10 && < 18.

Reply:
I updated the way to output the trailer. It incrementally outputs the trailer like the header now. So even the buffer size is less than 8, it also works. Added a test to `testSequenceFileGzipCodec` with block = 5.